external-plugins/rehearse/plugin/handler/handler.go

```diff
@@ -0,0 +1,463 @@
+
+	if !h.shouldActOnPREvent(event) {
+		log.Infoln("Skipping event. Not of our interest.")
+		return
+	}
+
+	if !h.validateEventUser(event.Repo.FullName, event.Sender.Login, event.PullRequest.Number) {
+		log.Infoln("Skipping event. User is not authorized.")
+		return
+	}
+
+	h.handleRehearsalForPR(log, &event.PullRequest, event.GUID)
+}
+
+func (h *GitHubEventsHandler) handleRehearsalForPR(log *logrus.Entry, pr *github.PullRequest, eventGUID string) {
+	repo, org, err := gitv2.OrgRepo(pr.Head.Repo.FullName)
+	if err != nil {
+		log.WithError(err).Errorf("Could not parse repo name: %s", pr.Head.Repo.FullName)
+		return
+	}
+	log.Infoln("Generating git client")
+	git, err := h.gitClientFactory.ClientFor(repo, org)
+	if err != nil {
+		return
+	}
+
+	log.Infoln("Fetching PR base ref", pr.Base.SHA)
+	err = git.FetchRef(pr.Base.SHA)
+	if err != nil {
+		log.WithError(err).Error("Could not fetch pull request's base ref.")
+		return
+	}
+	log.Infoln("Fetching PR head ref", pr.Head.SHA)
+	err = git.FetchRef(pr.Head.SHA)
+	if err != nil {
+		log.WithError(err).Error("Could not fetch pull request's head ref.")
+		return
+	}
+	log.Infoln("Getting diff")
+	changedFiles, err := git.Diff(pr.Head.SHA, pr.Base.SHA)
+	if err != nil {
+		log.WithError(err).Error("Could not calculate diff for PR.")
+		return
+	}
+	log.Infoln("Changed files:", changedFiles)
+	changedJobConfigs, err := h.modifiedJobConfigs(changedFiles)
+	if err != nil {
+		log.WithError(err).Error("Could not calculate absolute paths for modified job configs")
+		return
+	}
+	log.Infoln("Changed job configs:", changedJobConfigs)
+	headConfigs, err := h.loadConfigsAtRef(changedJobConfigs, git, pr.Head.SHA)
+	if err != nil {
+		log.WithError(err).Errorf(
+			"Could not load job configs from head ref: %s", pr.Head.SHA)
+	}
+
+	baseConfigs, err := h.loadConfigsAtRef(changedJobConfigs, git, pr.Base.SHA)
+	if err != nil {
+		log.WithError(err).Errorf(
+			"Could not load job configs from base ref: %s", pr.Base.SHA)
+	}
+	log.Infoln("Base configs:", baseConfigs)
+
+	prowjobs := h.generateProwJobs(headConfigs, baseConfigs, pr, eventGUID)
+	log.Infof("Will create %d jobs", len(prowjobs))
+	for _, job := range prowjobs {
+		if job.Labels == nil {
+			job.Labels = make(map[string]string)
+		}
+
+		if !shouldActOnJob(job) {
+			h.logger.Infof("Skipping rehearsal for job %s because it is not permitted", job.Name)
+			continue
+		}
+
+		job.Labels["rehearsal"] = "true"
+		job.Labels["rehearsal-for-pull-request"] = strconv.Itoa(pr.Number)
+		rehearsalName := strings.Join([]string{"rehearsal", job.Spec.Job}, "-")
+		job.Spec.Job = rehearsalName
+		job.Spec.Context = rehearsalName
+		_, err := h.prowClient.Create(&job)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to create prow job: %s", job.Spec.Job)
+		}
+		log.Infof("Created a rehearse job: %s", job.Name)
+	}
+}
+
+const rehearsalPermittedLabel = "rehearsal.allowed"
```

Comment:
Can you make this an annotation? Labels are not really a good fit for something like this.

Reply:
For presubmits I think we only have jobs in kubevirtci which would require blacklisting. I wonder if we should start with a denylist instead of an allowlist.

Reply:
We currently don't rehearse postsubmits and periodics. I think that it should be ok to start with explicit "allowed" annotation. Otherwise, someone can add a job by mistake that shouldn't be rehearsed and it will run.

Let's start with that and if it will become an issue, I'll update the code?

Reply:
Yes, since there are no postsubmits and periodics we basically add to every presubmit an allowed annotation. The only presubmits where we would maybe deny it are the push jobs from kubevirtci. It sounds better to me to explicitly deny the push jobs (if we we even want that. They "release" something but it is not referenced anywhere, so it is not dangerous).